### ISSUE TITLE
Endless queues with auto-generated names keep being created

### ISSUE THREAD

**Issue by mitchellporter** [hunterloftis/jackrabbit#56]
I have a singleton clock process that needs to publish some messages to workers once every 10 - 30 seconds. This works fine, but for some reason I end up generating a ton of queues, all with auto generated / random names. Here's a screenshot of the "Queues" tab in my `rabbitmq` dashboard after letting the clock process run for a few minutes:

<img width="904" alt="screen shot 2016-10-05 at 1 28 43 pm" src="https://cloud.githubusercontent.com/assets/7875419/19130780/98b137c8-8b01-11e6-8627-fa7e5e961e8d.png">

Below these queues are my real ones that I am creating, you just can't see them in the screenshot. I should only have about 10 of them. I have no idea why these other queues are being created or what they are being used for.

Here is my `jackrabbit` code that causes this:

```
var exchange = rabbit.default();
exchange.queue({ name: MY_QUEUE_NAME, durable: true });
exchange.publish({ update_info: 'some data' }, { key: MY_QUEUE_NAME });
```

If I add `exchange.on('drain', rabbit.close);` to the end of the above code it will solve this issue, but then I run into another issue which I posted here: https://github.com/hunterloftis/jackrabbit/issues/55

I'm not sure what I'm doing wrong here and would appreciate any feedback, help, or advice that anyone could offer. Thanks.

**Comment by Akamaozu** [comment]
Looks like every time for worker starts up, it creates a new exchange and as many queues as you need to go with it.

What you should do to solve the problem is define an exchange name for the worker. 

That way, it will create the exchange if it doesn't exist, or connect to it if it does.

**Comment by mitchellporter** [comment]
Hey @Akamaozu,

Here some example code from my worker.js:

```
var WORKERS = config.worker_concurrency;

throng({
    workers: WORKERS,
    lifetime: Infinity
}, start);

function start() {
 
 // Default exchange 
 var exchange = rabbit.default();
    
// Each one of these are my different types of jobs
 var create_broadcast = exchange.queue({ name: CREATE_BROADCAST_QUEUE, durable: true });
 var update_broadcast_processing = exchange.queue({ name: UPDATE_BROADCAST_PROCESSING_QUEUE, durable: true });
 var update_broadcast_vod = exchange.queue({ name: UPDATE_BROADCAST_VOD_QUEUE, durable: true });
 var delete_broadcast = exchange.queue({ name: DELETE_BROADCAST_QUEUE, durable: true });
 var sort_live_feed = exchange.queue({ name: SORT_LIVE_FEED_QUEUE, durable: true });
 var sort_vod_feed = exchange.queue({ name: SORT_VOD_FEED_QUEUE, durable: true });
 var sort_suggested_users = exchange.queue({ name: SORT_SUGGESTED_USERS_QUEUE, durable: true });
 var update_user = exchange.queue({ name: UPDATE_USER_QUEUE, durable: true });

// Handler functions for each type of job
 create_broadcast.consume(createBroadcast);
 update_broadcast_processing.consume(updateBroadcastProcessing);
 update_broadcast_vod.consume(updateBroadcastVOD);
 delete_broadcast.consume(deleteBroadcast);
 sort_live_feed.consume(sortFeed);
 sort_vod_feed.consume(sortFeed);
 sort_suggested_users.consume(sortSuggestedUsers);
 update_user.consume(updateUser);

}
````

And then here is an example of one of the handler functions:

```
function createBroadcast(data, ack) {
   // Do a bunch of stuff with the data.....

   // Ack the message
   ack();
}

```

Do you see anything that is wrong with this or do you have any suggestions on how I should change this?

Thanks for the help.

**Comment by mitchellporter** [comment]
After responding to you @Akamaozu I think I figured it out.

I'm going to copy over some of what I posted in my other issue #55:

Basically here's what I need.. I have multiple cron jobs that run every 10 to 30 seconds. I need to be able to send a message/job to my workers depending on which cron just fired. So here's an example of me creating my recurring cron jobs inside of my `clock.js`:

```
new CronJob('*/10 * * * * *', sortLiveFeed, null, true, 'America/Los_Angeles');
new CronJob('*/15 * * * * *', sortVODFeed, null, true, 'America/Los_Angeles');

```

The cron jobs above call these functions:

```
function sortLiveFeed() {
    
    // Publish message to queue
    var exchange = rabbit.default();
    exchange.queue({ name: SORT_LIVE_FEED_QUEUE, durable: true });
    exchange.publish({ status: 'live' }, { key: SORT_LIVE_FEED_QUEUE });
}
```

```
function sortVODFeed() {
    
    // Publish message to queue
    var exchange = rabbit.default();
    exchange.queue({ name: SORT_VOD_FEED_QUEUE, durable: true });
    exchange.publish({ status: 'vod' }, { key: SORT_VOD_FEED_QUEUE });
}
```

I basically just have a set of 10 different jobs that I will use throughout my entire app, and 4 of them happen to be created whenever their cron job fires. Above I showed you an example of 2 of these.

If you look above, both `sortLiveFeed()` and `sortVODFeed()` are called every 10-15 seconds. Well, in both functions, I am calling  `var exchange = rabbit.default();`, which I believe was causing the problem. I was creating a new exchange / queue? every single time these functions were being called.

I moved my exchange code and I'm now defining it only once at the top of the file and above the cron jobs code:

```
var exchange = rabbit.default();
new CronJob('*/10 * * * * *', sortLiveFeed, null, true, 'America/Los_Angeles');
new CronJob('*/15 * * * * *', sortVODFeed, null, true, 'America/Los_Angeles');
```

```
function sortLiveFeed() {
    
    // Publish message to queue
    exchange.queue({ name: SORT_LIVE_FEED_QUEUE, durable: true });
    exchange.publish({ status: 'live' }, { key: SORT_LIVE_FEED_QUEUE });
}

function sortVODFeed() {
    
    // Publish message to queue
    exchange.queue({ name: SORT_VOD_FEED_QUEUE, durable: true });
    exchange.publish({ status: 'vod' }, { key: SORT_VOD_FEED_QUEUE });
}
```

That solves the problem for me. I don't have endless queues being created and showing up in my `rabbitmq` dashboard. I would love to hear your feedback on this and I'm still open to any advice you can offer.

Thanks.

**Comment by Akamaozu** [comment]
When you use `rabbit.default()`, you are creating an unnamed exchange and attaching queues to it.

If you did `rabbit.default( 'sort_vod' )`, you'd be creating a named exchange, which will refer to an existing exchange if one exists, as opposed to creating new ones.

You've solved the problem in your context, but I'd strongly recommend named exchanges to avoid the same issue in future.

**Comment by mitchellporter** [comment]
>You've solved the problem in your context, but I'd strongly recommend named exchanges to avoid the same issue in future.

Are named exchanges useful in task / worker queue rabbitmq use cases? Is there a best practice, like for example should I create one named exchange called `worker_exchange`?

**Comment by Akamaozu** [comment]
Seeing this for the first time three years later 🙃

Hope you figured out the benefits of named exchanges by now.

Names allow my workers and producers to work together without having to do the hard work of making sure they all connect to the same exchange.

**Comment by mitchellporter** [comment]
@Akamaozu Haha no worries! If I remember correctly I ended up taking your advice and switching to named exchanges. Was funny reading these comments because I was so deep in the problem and burnt out at the time that I don't think I realized what you were trying to say and how easy the fix was.

Cheers!
